src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -350,7 +353,18 @@ internal MappedField GetOrAddDataField(ImmutableArray<byte> data, ushort alignme
                     bytesToStringHelper: bytesToStringHelper,
                     diagnostics: diagnostics);
             },
-            (@this, ImmutableCollectionsMarshal.AsImmutableArray(data), diagnostics)).Field;
+            (@this, ImmutableCollectionsMarshal.AsImmutableArray(data), diagnostics));
+
+            var previousText = @this._dataSectionStringLiteralNames.GetOrAdd(holder.Name, text);
+            if (previousText != text)
+            {
+                // If there is a hash collision, we cannot fallback to normal string literal emit strategy
+                // because the selection of which literal would get which emit strategy would not be deterministic.
+                var messageProvider = @this.ModuleBuilder.CommonCompilation.MessageProvider;
+                diagnostics.Add(messageProvider.CreateDiagnostic(messageProvider.ERR_DataSectionStringLiteralHashCollision, syntaxNode.GetLocation(), previousText));
+            }
```

Comment:
Curious whether this block could go inside the _dataSectionStringLiteralTypes.GetOrAdd call.